src/ucx_uct_plugin.c

```diff
@@ -0,0 +1,1613 @@
+  uct_iface_h iface           = NULL;
+  uct_component_h *comps, *comp;
+  unsigned comps_count, i;
+  ucs_status_t status;
+  uct_component_attr_t comp_attr;
+  uct_iface_attr_t iface_attr;
+  uct_md_h md;
+  uct_md_attr_t md_attr;
+  int rtr_size;
+
+  status = uct_query_components(&comps, &comps_count);
+  if (status != UCS_OK) {
+    WARN("Failed to query component list: error %d", status);
+    return NULL;
+  }
+
+  for (comp = comps; comp < comps + comps_count; comp++) {
+    comp_attr.field_mask = UCT_COMPONENT_ATTR_FIELD_MD_RESOURCE_COUNT;
+    status               = uct_component_query(*comp, &comp_attr);
+    if (status != UCS_OK) {
+      WARN("Failed to query component: error %d", status);
+      goto out;
+    }
+
+    comp_attr.field_mask = UCT_COMPONENT_ATTR_FIELD_MD_RESOURCES;
+    comp_attr.md_resources =
+        alloca(sizeof(*comp_attr.md_resources) * comp_attr.md_resource_count);
+    status = uct_component_query(*comp, &comp_attr);
+    if (status != UCS_OK) {
+      WARN("Failed to query component resources: error %d", status);
+      goto out;
+    }
+
+    for (i = 0; i < comp_attr.md_resource_count; i++) {
+      iface = nccl_uct_md_iface_open(worker, *comp, i,
+                                     comp_attr.md_resources[i].md_name, tl_name,
+                                     dev_name, &md);
+      if (iface != NULL) {
+        goto found;
+      }
+    }
+  }
+
+  if (iface == NULL) {
+    WARN("Failed to open iface for tl_name=%s dev_name=%s", tl_name, dev_name);
+    goto out;
+  }
+
+found:
+  status = uct_iface_query(iface, &iface_attr);
+  if (status != UCS_OK) {
+    WARN("Failed to query iface for tl_name=%s dev_name=%s", tl_name, dev_name);
+    goto fail;
+  }
+
+  if (!(iface_attr.cap.flags & UCT_IFACE_FLAG_CONNECT_TO_EP)) {
+    WARN("Interface flag CONNECT_TO_EP is not set");
+    goto fail;
+  }
+
+  if (!(iface_attr.cap.flags & UCT_IFACE_FLAG_GET_ZCOPY) ||
+      !(iface_attr.cap.flags & UCT_IFACE_FLAG_PUT_ZCOPY)) {
+    WARN("Interface does not support ZCOPY (flags=0x%x)", iface_attr.cap.flags);
+    goto fail;
+  }
+
+  status = uct_md_query(md, &md_attr);
+  if (status != UCS_OK) {
+    WARN("Failed to query md for iface %p: error %d", iface, status);
+    goto fail;
+  }
+
+  uct_iface = calloc(1, sizeof(*uct_iface));
+  if (uct_iface == NULL) {
+    WARN("Failed to alloc uct iface structure");
+    goto fail;
+  }
+
+  uct_iface->ep_addr_size     = iface_attr.ep_addr_len;
+  uct_iface->md               = md;
+  uct_iface->comp             = *comp;
+  uct_iface->rkey_packed_size = md_attr.rkey_packed_size;
+
+  if (iface_attr.cap.flags & UCT_IFACE_FLAG_AM_SHORT) {
+    uct_iface->am_max_short = iface_attr.cap.am.max_short;
+  }
+
+  rtr_size = nccl_uct_rdesc_size(NCCL_UCX_UCT_MAX_RECVS);
+  if (rtr_size > uct_iface->am_max_short) {
+    WARN("Failed RTR does not fit in face AM short (%dB > %dB)", rtr_size,
+         uct_iface->am_max_short);
+    goto fail;
+  }
```

Comment:
we can implement a fallback to `am_bcopy` if `max_am_short` is not big enough

Reply:
seems `max_am_short` is much bigger (2kB), so I'd keep that change for later.

Reply:
ok